sql/sem/tree: type unknown literals as `BPCHAR` if overload exists

mgartner opened the pull request:
#### sql/types: export BPCHAR type

The `typeBpChar` type, which is a `CHAR` type with unspecific length, is
now exported as `BPChar` for future use outside of the types package.

Release note: None

#### sql/logictest: move non-regression test above regression tests

Release note: None

#### sql/sem/tree: add third level of overload preference

Overload preference can now be one of three values—preferred, none, and
unpreferred—instead of two—preferred, and not preferred. The unpreferred
value is necessary in order to allow defining additional operator
overloads with input types that are equivalent (see
`(*types.T).Equivalent`) to input types of existing overloads without
causing statements to result in "ambiguous overload" errors.

An integer is used for overload preference, which allows for defining
addition levels of preference in the future, should those be necessary.

Release note: None

#### sql/sem/tree: type unknown literals as `BPCHAR` if overload exists

String literals in binary expressions can now be typed as a `BPCHAR` if
the other input to the binary expression has type `BPCHAR` and an
overload exists specifically with the `BPCHAR` type. When the literal is
typed as a `BPCHAR`, any trailing whitespace in the literal becomes
semantically insignificant.

Fixes #132268
Informs #75101

Release note (bug fix): A bug has been fixed that caused incorrect
evaluation of some binary expressions involving `CHAR(N)` values and
untyped string literals with trailing whitespace characters. For
example, the expression `'f'::CHAR = 'f '` now correctly evaluates to
`true`.

mgartner committed:
sql/types: export BPCHAR type

The `typeBpChar` type, which is a `CHAR` type with unspecific length, is
now exported as `BPChar` for future use outside of the types package.

Release note: None

mgartner committed:
sql/logictest: move non-regression test above regression tests

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133037)
<!-- Reviewable:end -->

mgartner committed:
sql/sem/tree: add third level of overload preference

Overload preference can now be one of three values—preferred, none, and
unpreferred—instead of two—preferred, and not preferred. The unpreferred
value is necessary in order to allow defining additional operator
overloads with input types that are equivalent (see
`(*types.T).Equivalent`) to input types of existing overloads without
causing statements to result in "ambiguous overload" errors.

An integer is used for overload preference, which allows for defining
addition levels of preference in the future, should those be necessary.

Release note: None

mgartner committed:
sql/sem/tree: type unknown literals as `BPCHAR` if overload exists

String literals in binary expressions can now be typed as a `BPCHAR` if
the other input to the binary expression has type `BPCHAR` and an
overload exists specifically with the `BPCHAR` type. When the literal is
typed as a `BPCHAR`, any trailing whitespace in the literal becomes
semantically insignificant.

Fixes #132268
Informs #75101

Release note (bug fix): A bug has been fixed that caused incorrect
evaluation of some binary expressions involving `CHAR(N)` values and
untyped string literals with trailing whitespace characters. For
example, the expression `'f'::CHAR = 'f '` now correctly evaluates to
`true`.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice fix!

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 10 of 10 files at r3, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-:-OAZkSEyDf5fjWUjKLcw:b-46lt7k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @asg0451, @mgartner, and @mw5h)

___
*[`pkg/config/system.go` line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-OAZ7TFH3oIeGubV1q60:-OAZ7TFH3oIeGubV1q61:b-4v5s0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/923483ae81549af7dea9500d9f86ebb8f14e91ae/pkg/config/system.go#L58)):*
> ```Go
> 
> 	// combined merges the zone and placeholder configs into a combined config.
> 	// If both have subzone information, the placeholder information is preference.
> ```

[nit] Looks like accidental match-and-replace.
___
*[`pkg/sql/logictest/testdata/logic_test/typing` line 228 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-OAZjznEC54xFojXoMjI:-OAZjznEC54xFojXoMjJ:bdndrso) ([raw file](https://github.com/cockroachdb/cockroach/blob/be336b83c82e064e48a36c2eb6448aa4f1809df1/pkg/sql/logictest/testdata/logic_test/typing#L228)):*
> ```
> 
> # The RHS is typed as CHAR(1) and truncated and the LHS is typed as BPCHAR, so
> # they do not match.
> ```

[nit] Isn't it the LHS that's typed as CHAR(1)?
___
*[`pkg/sql/logictest/testdata/logic_test/typing` line 282 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-OAZjQ_G3KPrCpt-qduM:-OAZjQ_HCTZrLJi_bRxi:b-34houj) ([raw file](https://github.com/cockroachdb/cockroach/blob/be336b83c82e064e48a36c2eb6448aa4f1809df1/pkg/sql/logictest/testdata/logic_test/typing#L282)):*
> ```
> ----
> false
> ```

What happens for this test?
```
postgres=# SELECT ROW('foo'::BPCHAR) = ROW('foo   ');
 ?column?
----------
 t
(1 row)
```


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133037)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @asg0451, @DrewKimball, and @mw5h)

___
*[`pkg/config/system.go` line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-OAZ7TFH3oIeGubV1q60:-OAc-DztAb-zuvLvxA7g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/923483ae81549af7dea9500d9f86ebb8f14e91ae/pkg/config/system.go#L58)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Looks like accidental match-and-replace.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/typing` line 228 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-OAZjznEC54xFojXoMjI:-OAc-NhC9TDrw_87JhEM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be336b83c82e064e48a36c2eb6448aa4f1809df1/pkg/sql/logictest/testdata/logic_test/typing#L228)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Isn't it the LHS that's typed as CHAR(1)?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/typing` line 282 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133037#-OAZjQ_G3KPrCpt-qduM:-OAc-tI44yjz2sMOT8TF:b-t26iu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/be336b83c82e064e48a36c2eb6448aa4f1809df1/pkg/sql/logictest/testdata/logic_test/typing#L282)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What happens for this test?
```
postgres=# SELECT ROW('foo'::BPCHAR) = ROW('foo   ');
 ?column?
----------
 t
(1 row)
```
</blockquote></details>

Good idea - I added this test. The correct result, `true`,  is returned. :tada: 


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
> Reviewed 17 of 17 files at r5, 1 of 1 files at r6, 10 of 10 files at r7, 10 of 10 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133037)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asg0451 and @mw5h)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+

mgartner commented:
Looks like a test infra flake:

<img width="1128" alt="image" src="https://github.com/user-attachments/assets/6c556b20-cc91-47f1-b49b-2c63fe4f84ed">

bors r+

mgartner commented:
blathers backport 24.3
